### ISSUE TITLE
compile_template expects unicode file, but converts it to plain string

### ISSUE THREAD

**Issue by damon-atkins** [saltstack/salt#36113]
### Description of Issue/Question

When a template is open, it is open with
codecs.open(template, encoding='utf-8')

In salt\template.py the following line is used
input_data = string_io(input_data)
Before this line input_data is unicode.
After this line of code its a StringIO object, but the contents is a string (Non-Unicode)

It seems an issue to open a file as utf-8 and then convert it to a string (Non-Unicode) before rendering it.
### Steps to Reproduce Issue

```
    log.error('DA input_data type is {0}'.format(type(input_data)))
    input_data = string_io(input_data)
    log.error('DA input_data type is {0}'.format(type(input_data)))
    DA=input_data.getvalue()
    log.error('DA input_data[DA] type is {0}'.format(type(DA)))
```

```
[ERROR   ] DA input_data type is <type 'unicode'>
[ERROR   ] DA input_data type is <type 'cStringIO.StringI'>
[ERROR   ] DA input_data[DA] type is <type 'str'>
```
### Versions Report

Develop on Windows and assume many older versions

Part of the fix might be:

```
#from salt._compat import string_io
from salt.ext.six.moves import StringIO # Don't use faster cStringIO no Unicode Support

input_data=StringIO(input_data)
```

However the renders seem to be returning strings if you provide for example a yaml state file.

**Comment by thatch45** [comment]
While I basically agree, this has been a bit of a hicup for us. Are you experiencing any issues right now due to this?

**Comment by damon-atkins** [comment]
I am surprise its not impacting windows packaging.   For example skype package sls contains the TM single utf-8 character symbol.  And then thier is software which will have international language name.  I have not tried this with PY3 to see what happens.   I suggest you add this to PY2 PY3 compatible check list.   template.py seems to be the only code calling string_io

At this stage I cannot point to something and say its **end result** is broken and this is the cause.

To test further for a impact need some software which is listed in windows uninstall with a non English name, then write a sls to install it.

(Any one has a simple free app with non English name listed in Uninstall, post it here pls. In fact it would be could to have some fake test apps for salt as part of unit testing)

**Comment by damon-atkins** [comment]
The following code writes out the winrepo.p file for windows packaging
```
    with salt.utils.fopen(winrepo, 'w+b') as repo_cache:
        repo_cache.write(msgpack.dumps(ret))
    #__context__['winrepo.data'] = ret  # save reading it back again
```
If I enable the line `__context__['winrepo.data'] = ret`  it breaks.    However if `__context__['winrepo.data'] ` is set after reading the file back in (see coded below) its fine.
```
    with salt.utils.fopen(winrepo, 'rb') as repofile:
        try:
            repodata = msgpack.loads(repofile.read()) or {}
            __context__['winrepo.data'] = repodata
```
The breaking error is
```
    File "C:\salt\bin\lib\site-packages\salt\modules\win_pkg.py", line 263, in list_pkgs
      name_map = _get_name_map(saltenv)
     File "C:\salt\bin\lib\site-packages\salt\modules\win_pkg.py", line 1498, in _get_name_map
       u_name_map[k.decode('utf-8')] = name_map[k]
     File "C:\salt\bin\lib\encodings\utf_8.py", line 16, in decode
       return codecs.utf_8_decode(input, errors, True)
  UnicodeEncodeError: 'ascii' codec can't encode character u'\u2122' in position 5: ordinal not in range(128)
```
msgpack.dumps or msgpack.loads  is altering the original data which just happens to make it work.

**Comment by damon-atkins** [comment]
This should be resolved in 2018.3
